REGRESSION(?): [iOS] ASSERTION FAILED: result in WebKit::NetworkProcess::allowsFirstPartyForCookies()

youennf opened the pull request:
#### 1b0d8f7b12c4b763fadd606974c5167284436989
<pre>
REGRESSION(?): [iOS] ASSERTION FAILED: result in WebKit::NetworkProcess::allowsFirstPartyForCookies()
<a href="https://bugs.webkit.org/show_bug.cgi?id=248065">https://bugs.webkit.org/show_bug.cgi?id=248065</a>
rdar://problem/102490719

Reviewed by Alex Christensen.

We may reuse a process for running shared workers with different first party cookie URLs.
We therefore need for each shared worker to make sure the NetworkProcess allowed first party URLs map is correctly populated.
We move the population at the place where we launch the shared worker instead of the time where the worker process connection is created.
This aligns with how this is done with service workers.

Covered by imported/w3c/web-platform-tests/service-workers/cache-storage/cross-partition.https.tentative.html no longer crashing.

* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::didFinishFetchingSharedWorkerScript):
(WebKit::WebSharedWorkerServer::createContextConnection):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.cpp:
(WebKit::WebSharedWorkerServerToContextConnection::launchSharedWorker):

Canonical link: <a href="https://commits.webkit.org/259417@main">https://commits.webkit.org/259417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43bac5d2642630614e037ad2b9e408a651be5073

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114132 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4867 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97190 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113157 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110623 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94649 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39164 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93511 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26267 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7287 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27628 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7383 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4221 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13439 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47178 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6496 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9172 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
```diff
@@ -145,22 +145,19 @@ void WebSharedWorkerServer::createContextConnection(const WebCore::RegistrableDo
 
         RELEASE_LOG(SharedWorker, "WebSharedWorkerServer::createContextConnection should now have created a connection");
 
-        // FIXME: Add a check that the process this firstPartyForCookies came from was allowed to use it as a firstPartyForCookies.
```
This comment should be kept.

youennf replied:
OK, let's move it to WebSharedWorkerServerConnection::requestSharedWorker, which seems to be the right place for it then.

youennf committed:
REGRESSION(?): [iOS] ASSERTION FAILED: result in WebKit::NetworkProcess::allowsFirstPartyForCookies()
https://bugs.webkit.org/show_bug.cgi?id=248065
rdar://problem/102490719

Reviewed by Alex Christensen.

We may reuse a process for running shared workers with different first party cookie URLs.
We therefore need for each shared worker to make sure the NetworkProcess allowed first party URLs map is correctly populated.
We move the population at the place where we launch the shared worker instead of the time where the worker process connection is created.
This aligns with how this is done with service workers.

Covered by imported/w3c/web-platform-tests/service-workers/cache-storage/cross-partition.https.tentative.html no longer crashing.

* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::didFinishFetchingSharedWorkerScript):
(WebKit::WebSharedWorkerServer::createContextConnection):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerToContextConnection.cpp:
(WebKit::WebSharedWorkerServerToContextConnection::launchSharedWorker):

Canonical link: https://commits.webkit.org/259417@main
